### DISCUSSION TITLE
Get the currently logged in Wiki.js username in javascript?

### DISCUSSION THREAD

**Discussion by rmunroe** [requarks/wiki#6583] [Help / Questions] [answered]
I am using the wiki for a private, internal knowledge base and need to track who is / is not accessing it. All users must be logged in, and we are using OIDC (Google) as the auth provider. We are using Matomo for click tracking, which allows for tracking by username via a javascript call. The only problem is getting the username from Wiki.js / OIDC / Google to sent to Matomo.

Is there any way to get the currently logged in Wiki.js username in javascript?

**Comment by NGPixel** [comment, marked as answer]
You can use these methods to get the info you want:

```
WIKI.$store.get('user/id')
WIKI.$store.get('user/name')
WIKI.$store.get('user/email')
```

However, you need to wait for the Vue app to be initialized first or they won't exist / return the correct data.

See https://docs.requarks.io/en/troubleshooting#how-to-execute-javascript-code-on-page-load

**Comment by NicksonYap** [comment]
I added the line `_paq.push(['setUserId', String(WIKI.$store.get('user/id'))]);` wrapped with `window.boot.register('page-ready', function() {});`

In `server/modules/analytics/matomo/code.yml`

```yml
head: |
  <!-- Matomo -->
  <script type="text/javascript">
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
      window.boot.register('page-ready', function() {
        var u="{{serverHost}}/";
        _paq.push(['setTrackerUrl', u+'matomo.php']);
        _paq.push(['setSiteId', '{{siteId}}']);

        //ref: https://matomo.org/faq/reports/set-up-user-id-tracking-in-matomo/
        //ref: https://github.com/requarks/wiki/discussions/5126
        //ref: https://github.com/requarks/wiki/discussions/6583
        //WIKI.$store.get('user/id')
        //WIKI.$store.get('user/name')
        //WIKI.$store.get('user/email')
        _paq.push(['setUserId', String(WIKI.$store.get('user/id'))]);

        var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
        g.type='text/javascript'; g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
      });
    })();
  </script>
  <noscript><p><img src="{{serverHost}}/matomo.php?idsite={{siteId}}&amp;rec=1" style="border:0;" alt="" /></p></noscript>
  <!-- End Matomo Code -->

```

**Reply by ICantCodeEtAl** [reply]
Man, so appreciated! Was looking to do this exact thing minus google and was failing massively. Well done! Just rebooted the server, hoping this works.

**Comment by NicksonYap** [comment]
Appreciate the pointer! @NGPixel

I did some tests, and above code is working when logged in, and in Matomo, it combine visits across devices for the same user. 
But, if user is logged out, then `page-ready` does not fire at all, causing the Matomo JS script to not trigger at all

This is no biggie as I do intend to track logged in Users for now, but I do intend to open some Wiki pages to the public, in the future then this implementation would not work>

Is there another event that will fire for non-logged in conditions?
Or was there something I did wrong above?

Best,
Nickson
